### DISCUSSION TITLE
Avalonia Windows ZOrder

### DISCUSSION THREAD

**Discussion by npolyak** [AvaloniaUI/Avalonia#6460] [Q&A] [answered]
Friends, is there a property to control and check Avalonia Window ZOrder across multiple platforms? A property that would specify which window is on top of another window and allow to change it. Window's ZIndex property does not seem to work at all. 

Looking forward to some help in solving this puzzle.

**Comment by npolyak** [comment]
Can some one answer this question?

**Comment by PieroCastillo** [comment]
I think u can do it requesting to the windowing system (dwm in windows, x11/wayland linux, and obviusly the ws of macOS)

**Reply by npolyak** [reply]
Can you give me the commands or a little more documentation - for lunix and MacOS? 

Perhaps there are other - cross-platform solutions?

Thanks

**Reply by PieroCastillo** [reply]
https://www.generacodice.com/en/articolo/71990/How-to-get-the-z-order-in-windows? https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-gettopwindow i think this can help u, i think u should get all processes, next check what processes are windowed (remember the existence of backgorund process, and finally make an algorithm to check the order of the windows)

**Reply by maxkatz6** [reply]
On macOS you can create native NSWindow object from handle (avaloniaWnd.PlatformImpl.Handle). It will give you access to the native windowing API. And there also will be ZOrder property that works with windows of the app (as far as I can tell, have never tried it before). 
And to use this NSWindow API you need to either use net6-mac target from NET 6 (haven't used it yet too, but should just work) or simply reference MonoMac-netstandard package from nuget.

**Reply by maxkatz6** [reply]
In general you can't avoid using system APIs sooner or later, so it worth to get familiar with it on all supported by you platforms. 
Avalonia can cover some aspects of it, but it's just GUI framework with limited responsibility (in this case I agree window ZIndex makes sense).

**Reply by npolyak** [reply]
Platform specific coding is like death and taxes - unavoidable but preferably can be pushed off as far as possible :)

**Comment by maxkatz6** [comment, marked as answer]
I don't think there is an API for that in Avalonia, but it sounds like reasonable feature request, if you will create an issue for that.

**Reply by npolyak** [reply]
I created https://github.com/AvaloniaUI/Avalonia/issues/6479 issue to cover this. Thanks

**Comment by tomlm** [comment]
I discovered that ZIndex doesn't seem to get applied correctly when doing window.Show() unless you do it delayed.

I think it has something to do with the rendering of the visual tree and eventing.  At any rate, I ended up doing this:
```cs
                Dispatcher.UIThread.InvokeAsync(() => _floatingContent.Show(VisualRoot as Window));
```

**Reply by thevortexcloud** [reply]
You probably need to wait until the window is ready by using the `Loaded` event. A lot of things don't work until that event fires.
